### PR TITLE
Refactor `update_files.py`

### PR DESCRIPTION
**Pull request by desilinguist**
- Refactor all of the core file updating code into a class called `FileUpdater` located in `rsmtool.test_utils.py`. This makes the code much more organized and also makes it much easier to write tests for it in a subsequent release. 
- The code now doesn't raise any exceptions. It just records all of the information when then `run()` method is called and prints everything out when `print_report()` is called. 
- The code is also fully modular in that nothing is hardcoded. You need to pass in the directory containing the tests as well as the directory containing the updated test outputs. In addition, it also automatically gets all of the `test_experiment_*.py` files from the tests directory and uses those. This will make it possible to use the _same_ script for RSMExtra without any changes. 
- Tests decorated with `@raises` which do not need to be updated are now skipped much more directly rather than relying on an indirect signal previously. 
- Sometimes you might want to name the file `test_experiment_X.py` but _not_ update its test data automatically. For example, the file `test_experiment_rsmcompare.py` does actually contain experiment tests but since they do not contain an output directory, they do not need to be updated (for now). However, this file does get picked up by `update_files.py` because of its name and then we get reports of missing directories. I solved this by simply adding an `_AUTO_UPDATE` variable to the top of all of the `test_experiment_*.py` files and setting it to `True` for all files except for `test_experiment_rsmcompare.py`. `FileUpdater.run()` now checks for this before actually processing a file. 

In short, this change should allow us to use the same script for RSMExtra without any changes and also _always_ print out a useful report for the data update process.

### CODE REVIEW

**Comment by desilinguist** [comment]
The coverage went down because it's including `rsmtool/test_utils.py` in the coverage which we don't really test at all right now. I will tell Travis to ignore it for now.

File: rsmtool/test_utils.py [hunk 1]
```diff
@@ -700,3 +706,242 @@ def check_subgroup_outputs(output_dir, experiment_id, subgroups, file_format='cs
...
+            "test_experiment_" and located in the `tests_directory` to find
+            the tests that are to be updated.
+        updated_outputs_directory (TYPE): Description
+        deleted_files (list): List of files deleted from `tests directory`.
+        updated_files (list): List of files added or updated for the tests
+            under `tests_directory`.
+        missing_or_empty_sources (list): List of source names whose corresponding
+            directories are either missing under `updated_outputs_directory` or
+            do exist but are empty.
+    """
```
Comment (jbiggsets) [inline rsmtool/test_utils.py]: Can we change these to `numpy`-style docstrings, rather than Google-style?

```
Attributes
----------
tests_directory : str
    Path to the directory containing the tests whose
    outputs are to be updated.
```

Comment (desilinguist) [reply rsmtool/test_utils.py]: Weird, I could have sworn I had set my `AutoDocstring` to numpy and not google since it works fine for methods. I will redo.

File: rsmtool/test_utils.py [hunk 2]
```diff
@@ -700,3 +706,242 @@ def check_subgroup_outputs(output_dir, experiment_id, subgroups, file_format='cs
...
+        suffix = Path(filename).suffix
+        return suffix == '.model' or \
+            suffix == '.npy'or \
+            stem.endswith('_postprocessing_params') or \
+            stem.endswith('_eval') or \
+            stem.endswith('_eval_short') or \
+            stem.endswith('_confMatrix') or \
+            stem.endswith('_pred_train') or \
+            stem.endswith('_pred_processed') or \
+            stem.endswith('_score_dist')
```
Comment (jbiggsets) [inline rsmtool/test_utils.py]: This is probably not a big deal, but would we ever be adding to the list of possible `skll` suffixes or stems here? Is it easier to put these into lists, rather than having so many `or` statements?

```
possible_suffixes = ['.model', '.npy']
possible_stems = ['_postprocessing_params', '_eval', '_eval_short', ...]

return any(suffix == s for s in possible_suffixes) or \
    any(stem.endswith(s) for s in possible_stems)
```

File: rsmtool/test_utils.py [hunk 3]
```diff
@@ -700,3 +706,242 @@ def check_subgroup_outputs(output_dir, experiment_id, subgroups, file_format='cs
...
+        # since those are likely old files from old versions that we do not need
+        existing_output_only_files = dir_comparison.right_only
+        for file in existing_output_only_files:
+            remove(existing_output_path / file)
+
+        # Next find all the NEW files in the updated outputs but do not include
+        # the config JSON files or the OLS summary files or
+        # the evaluation/prediction/model files for SKLL models
+        new_files = dir_comparison.left_only
+        new_files = [f for f in new_files if not f.endswith('_rsmtool.json') and not f.endswith('_rsmeval.json') and not f.endswith('_ols_summary.txt')]
```
Comment (jbiggsets) [inline rsmtool/test_utils.py]: Again, would it be better to just use `any()` here and a list of filename suffixes, in case we expand on this in the future?

File: rsmtool/test_utils.py (hunk 1 above)
Comment (desilinguist) [reply rsmtool/test_utils.py]: Ah, it was set to `auto_google`, which means it tries to autodetect the style based on the other docstrings but falls back to google style if it can't.

File: tests/update_files.py [hunk 4]
```diff
@@ -157,90 +57,25 @@ def main():
     elif run_test_suite != 'y':
         print('Invalid answer. Exiting.')
         sys.exit(1)
+    else:
+        print()
+
+    # iterate over the given tests directory and find all files named
+    # `test_experiment_*.py` and get their suffixes for use with the
+    # FileUpdater object.
+    suffixes = [re.sub(r'test_experiment_', '', p.stem)
```
Comment (jbiggsets) [inline tests/update_files.py]: Is it really necessary to have the `_AUTO_UPDATE` global variables in each test module. Can't we just ignore `rsmcompare` here?

File: rsmtool/test_utils.py [hunk 5]
```diff
@@ -700,3 +706,242 @@ def check_subgroup_outputs(output_dir, experiment_id, subgroups, file_format='cs
...
+                    function = member_object
+
+                    # get the qualified name of the member function
+                    member_qualified_name = member_object.__qualname__
+
+                    # check if it has 'raises' in the qualified name and skip it, if it does
+                    if 'raises' in member_qualified_name:
+                        continue
+
+                    print(member_name)
```
Comment (jbiggsets) [inline rsmtool/test_utils.py]: Do we need this print here?

File: tests/update_files.py (hunk 4 above)
Comment (desilinguist) [reply tests/update_files.py]: It goes back to your other comment for the other things: what if we want to add more modules in the future we don't want to update? I am happy to hardcode it if you guys prefer.

File: rsmtool/test_utils.py (hunk 5 above)
Comment (desilinguist) [reply rsmtool/test_utils.py]: No, we don't :)

File: tests/update_files.py (hunk 4 above)
Comment (jbiggsets) [reply tests/update_files.py]: No, that makes sense. 

Is there somewhere we can document the we have to add `_AUTO_UPDATE` at the top if/when we add new `test_experiment_*.py` files? I'm just worried I'll totally forget about it. :)

Comment (desilinguist) [reply tests/update_files.py]: Yes, it will be in the testing document that I will be writing.

File: rsmtool/test_utils.py [hunk 6]
```diff
@@ -700,3 +706,251 @@ def check_subgroup_outputs(output_dir, experiment_id, subgroups, file_format='cs
...
+
+        # if the directory for this source does not exist on the updated output
+        # side, then that's a problem and something we should report on later
+        try:
+            assert updated_output_path.exists()
+        except AssertionError:
+            self.missing_or_empty_sources.append(source)
+            return
+
+        # if the updated output path does not exist, then create it
```
Comment (aloukina) [inline rsmtool/test_utils.py]: updated -> existing

File: rsmtool/test_utils.py [hunk 7]
```diff
@@ -700,3 +706,251 @@ def check_subgroup_outputs(output_dir, experiment_id, subgroups, file_format='cs
...
+                except AssertionError:
+                    pass
+                else:
+                    include_file = False
+
+            if include_file:
+                really_changed_files.append(changed_file)
+
+        # Copy over the new files as well as the really changed files
+        new_or_changed_files = new_files + really_changed_files
```
Comment (aloukina) [inline rsmtool/test_utils.py]: Why don't we  then have `self.updated_files` and `self.new_files`?

File: tests/update_files.py [hunk 8]
```diff
@@ -28,126 +29,25 @@
...
+from rsmtool.test_utils import FileUpdater
 
 
 def main():
     # set up an argument parser
     parser = argparse.ArgumentParser(prog='update_test_files.py')
-    parser.add_argument('outputs_dir', help="The path to the directory containing the updated test outputs")
+    parser.add_argument('--tests',
+                        dest='tests_dir',
+                        help="The path to the RSMTool tests directory")
```
Comment (aloukina) [inline tests/update_files.py]: 'existing'?

File: tests/update_files.py [hunk 9]
```diff
@@ -28,126 +29,25 @@
...
+    parser.add_argument('--outputs',
+                        dest='outputs_dir',
+                        help="The path to the directory containing the test "
+                             "outputs (usually test_outputs)")
```
Comment (aloukina) [inline tests/update_files.py]: 'updated'?

File: tests/update_files.py (hunk 4 above)
Comment (aloukina) [reply tests/update_files.py]: Will you also add to the documentation that the test script should be named `test_experiment_*`?

Comment (desilinguist) [reply tests/update_files.py]: Yes it’s in there already.

File: rsmtool/test_utils.py (hunk 7 above)
Comment (desilinguist) [reply rsmtool/test_utils.py]: You mean ALSO have `self.new_files` In addition to `self.updated_files`?

File: tests/update_files.py [hunk 10]
```diff
@@ -28,126 +29,25 @@
...
-    # return lists of files we deleted and updated
-    return sorted(existing_output_only_files), sorted(new_or_changed_files)
+from rsmtool.test_utils import FileUpdater
 
 
 def main():
     # set up an argument parser
     parser = argparse.ArgumentParser(prog='update_test_files.py')
-    parser.add_argument('outputs_dir', help="The path to the directory containing the updated test outputs")
+    parser.add_argument('--tests',
```
Comment (aloukina) [inline tests/update_files.py]: Potentially a nitpick: these are treated as optional arguments when you call the script so you get a Traceback error (Path is None) if you forget to specify them.

File: rsmtool/test_utils.py (hunk 7 above)
Comment (aloukina) [reply rsmtool/test_utils.py]: Yes: `self.new_files = new files` and `self.updated_files = `really_changed_files`. I don't feel too strongly about it though.

Comment (desilinguist) [reply rsmtool/test_utils.py]: That's a good idea. I can do that.

File: tests/update_files.py (hunk 10 above)
Comment (desilinguist) [reply tests/update_files.py]: I can just add `required=True` to these.

File: rsmtool/test_utils.py (hunk 7 above)
Comment (desilinguist) [reply rsmtool/test_utils.py]: Actually, I'd rather not do that since that requires a lot of changes all over the code.

Comment (aloukina) [reply rsmtool/test_utils.py]: That's fine - I'll file an issue for when we feel like it in future
